Add soft wolfe condition with tests

kumrud opened the pull request.

kumrud committed:
Add soft-wolfe (curvature) condition

1. Add sof-wolfe condition
2. Add tests

crisely09 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+        raise ValueError('Size of direction and variable should be same')
+    if not 0 < const2 < 1:
+        raise ValueError('C2 is not within boundary')
+    if not callable(gradf):
```
this is cool, I should use something like that too

crisely09 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+    const2: float
+        Constant C2
+
+    Raises
```
I've never thought of adding what errors raises, but I guess we should be consistent, so I'll add something similar then

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+    gradf: function
+        Gradient of objective nonlinear function
+    valf: np.ndarray(N,)
+        Function value
```
Is this x? I'm just getting confused because `gradf` is a function and `valf` is not a function.

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+    Parameters
+    ----------
+    gradf: function
+        Gradient of objective nonlinear function
```
It would be better to use `callable` instead of `function`, I think. Also, it would be good to include the input and the output of the function somewhere.

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+    Returns
+    -------
+    Boolean
+        True if curvature condition is satisfied
```
I'm not sure if `Boolean` is necessary. Could you also include `False otherwise`?

kimt33 replied:
Nevermind, the type would be necessary, but can it be changed to `bool` instead?

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+    Returns
+    -------
+    Boolean
+        True if curvature condition is satisfied
```
We probably need references for the documentation.

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+       raise TypeError('Descent direction should be an array of float')
+    if not isinstance(alpha, float):
+        raise TypeError("Alpha should be a float")
+    if xk.size != sk.size:
```
What are `xk` and `sk`? I didn't see them defined anywhere.

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+    if not isinstance(direction, (np.ndarray, float)):
+       raise TypeError('Descent direction should be an array of float')
+    if not isinstance(alpha, float):
+        raise TypeError("Alpha should be a float")
```
Sorry to nitpick, but single quote? (This isn't really a PEP8 thing or anything... but it does change from single quote to double quote here though)

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+        raise TypeError('Variable should be an array of float')
+    if not isinstance(direction, (np.ndarray, float)):
+       raise TypeError('Descent direction should be an array of float')
+    if not isinstance(alpha, float):
```
FYI we can also allow `np.ndarray` here to allow vector alpha (though it probably should not be supported yet)

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+    """
+    # Check parameters
+    if not isinstance(valf, (np.ndarray, float)):
+        raise TypeError('Variable should be an array of float')
```
If you want it to be an array of float, then you also need to check its data type:
```python
if valf.dtype != float:
    raise TypeError('wrong type')
```
I'm not sure if both float and numpy array should be supported because we might get some errors if we do not explicitly test for float input. For example, line 37 will fail if a float is given (because a float does not have attribute `size`). If it will be supported, then it should be converted into a numpy array for consistency within the code:
```python
if isinstance(valf, float):
    valf = np.array(valf)
elif not isinstance(valf, np.ndarray):
    raise TypeError
```

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+        raise TypeError("Alpha should be a float")
+    if xk.size != sk.size:
+        raise ValueError('Size of direction and variable should be same')
+    if not 0 < const2 < 1:
```
we should first check that `const2` is a float. If it is a numpy array, the code will raise an error here (because you cannot use `not` operator with a numpy array).

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
+    left_con = direction.dot(gradf(valf + alpha * direction))
+    right_con = const2 * direction.dot(gradf(valf))
+
+    return bool(left_con >= right_con)
```
the `bool` operator is not necessary

kimt33 commented on flik/linesearch/test/test_conditions.py:
```diff
@@ -1 +1,31 @@
+    """ Test soft wolfe/curvature condition."""
+    #set up parameters
+    def gradf(val):
+        return 2*val
```
short documentation here. e.g.
```python
"""Gradient of function :math:`x^2`."""
```
You don't have to use LaTeX, but we might have to learn to use it in the future (to document some of the more complex things)

kimt33 commented on flik/linesearch/test/test_conditions.py:
```diff
@@ -1 +1,31 @@
+
+    # Check condition
+    assert soft_wolfe(gradf, valf, alpha, direction)
+    assert soft_wolfe(gradf, valf, alpha, -direction) is False
```
use `not soft_wolfe(gradf, valf, alpha, -direction)` instead of `soft_wolfe(gradf, valf, alpha, -direction) is False`

kimt33 replied:
There should probably be more tests. We can use a variety of different functions, but we should probably test different directions, alpha, starting point, and constant. It'd be great if we can use values that are conceptually easy to understand. Maybe there's an example in the textbook.

Not too important right now, but we will need to address this issue when we demo this.

kimt33 commented on flik/linesearch/conditions.py:
```diff
@@ -1 +1,47 @@
 """Conditions for the line search algorithm."""
+def soft_wolfe(gradf, valf, alpha, direction, const2=0.9):
+    """Curvature condition
```
Soft Wolfe condition (or curvature condition)

kimt33 reviewed (changes requested):
I was just being picky about some things. I have a feeling that there is a commit that wasn't pushed (because the variable names are not consistent).

Either way, there are two "important" issues that I think we need to address. First, we should probably use type hinting. I'm not too familiar with it myself, but that's where the language is going so it's best to get accustomed to it early on. Second, we might need to use LaTeX to document some of the code. Right now, the general consensus is to forgo documenting equations and simply pointing to a source. I think that will have its limitations, and we may end up writing documentation for some of these things, which will require the use of LaTeX for equations. Not too sure if it is necessary right now, though.

kimt33 committed:
Clean up soft wolfe

1. Merge strong wolfe and soft wolfe as a single function, distinguished
   from one another with a keyword argument
2. Merge tests for strong and soft wolfe
